[ext-doc] Extension Architecture rewrite (#4482)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/mv3/architecture-overview/index.md

layout: "layouts/doc-post.njk"
title: "Architecture overview"
date: 2012-09-18
updated: 2022-12-16
description: A high-level explanation of the architecture of Chrome Extensions.
subhead: A high-level explanation of the structure of a Chrome Extension.
---

## Overview {: #overview }

> A Chrome extension is composed of different parts. This page describes the structure of an extension, the role each part plays and how they work together. Implementation details are out of the scope of this article.

Author: The primary purpose of this rewrite was to narrow the scope of this overview to focus on structure and relation between extension parts. I moved implementation details to other pages that cover specific topics.

> A Chrome extension is composed of different parts. This page describes the structure of an extension, the role each part plays and how they work together. Implementation details are out of the scope of this article.

Author: I think you are right, it does sound a bit blunt! 😅 

Both suggestions sound good  💯 , I'll have to pick one.

If you are not familiar with Chrome extension development, we recommend first reading [Extensions 101][doc-ext-101] and [Development Basics][doc-dev-basics].

> ## The structure of a Chrome extension {: #arch }

Author: I made sure to match the original document's header ids, to no break existing bookmarks or links to these resources

### The manifest {: #manifest }

The manifest is the configuration file of a Chrome extension. It is a required JSON file that must be located at the root of the project. It provides the browser with a blueprint of the extension, with important information such as:

- The name of the extension, a description of what it does, the current version number, and what icons to use.
- The [Chrome APIs][api-ref] keys and [permissions][doc-perms] that the extension needs.
- The files assigned as the extension service worker, the popup HTML file, the options page, the content scripts, etc.

The [Manifest keys][doc-manifest] article contains the complete list of default and optional keys. For copy-paste-ready code samples, check out the [Manifest examples][doc-manifest-examples].

### The service worker {: #background_script }

The extension service worker is the event manager of a Chrome extension. It is a script that runs in the background within the context of the browser. It monitors and responds to events in the browser. For example, when a new tab is created, a new bookmark is added, the extension is installed, the extension toolbar icon is clicked, etc.

> It can access all the [Chrome APIs][api-ref], but as a service worker, it can't use the DOM APIs that the global window object provides. It also cannot modify the page's content.

Reviewer: > but as a service worker

I like that this line line is trying to draw a distinction between a service worker and an _extension_ service worker, but I don't think readers are likely to pick up on that here. I expect that people reading this document will see this as saying "the extension's service worker can access Chrome APIs, but as a service worker, it can't use DOM APIs".  I think that reads as oddly self-reflective and doesn't really address the question of why.

> It also cannot modify the page's content.

I find this line a bit confusing. As a reader, my immediate question is "_the_ page? What page is that?"

It's good to highlight that the background script cannot modify pages (something that has historically been a stumbling block for new extension developers), but I think we can lean a little more into this idea without substantially increasing the complexity of this paragraph.

**Suggestion**

```suggestion
A service workers can access all the [Extension APIs][api-ref], but as a type of [Worker](https://developer.mozilla.org/en-US/docs/Web/API/Worker) it can't use the typical DOM APIs that a document's global Window object provides. It also runs in its own environment, so it cannot directly modify a web page's content.
```

See [Handling events in the extension service worker][doc-sw] for more details. 

### Content scripts {: #content-scripts }

> Extensions use content scripts to inject code into host pages. They allow the extension to interact with and modify the pages in the browser. For example, they can insert a new element on the page, change the style of a website, modify the [DOM][mdn-dom] elements, etc. 

Reviewer: ```suggestion
Extensions use content scripts to inject code into host pages. They allow the extension to interact with and modify pages in the browser. For example, they can insert a new element on the page, change the style of a website, modify the [DOM][mdn-dom] elements, etc. 
```

> Content Scripts share the same origin as their host page and they have access to a limited number of [Chrome APIs][api-ref]. See [Understanding content scripts][doc-content-scripts] for more details.

Reviewer: > Content Scripts share the same origin as their host page

While technically true (calling `new URL(window.location.href).origin` in a content script and directly in the page will give you the same value), I don't think this is the best thing to highlight here. Non-web developers likely won't have a sense of what this means or it's implications, and I expect web developers will mostly think of origins as the boundary for cookies or localStorage.

I think it would be better to focus this line on isolated worlds. 

```suggestion
Content Scripts share access to the same DOM tree as the host page but run in a separate JavaScript environment. They also have access to a limited number of [Chrome APIs][api-ref]. See [Understanding content scripts][doc-content-scripts] for more details.
```

### Extension HTML pages {: #html-files }

An extension can have different HTML pages depending on the design. All extension HTML files can use the [Chrome APIs][api-ref], but cannot include inline Javascript; they must point to a JavaScript file. The two most common HTML pages are:

[The popup][doc-popup]
: Many extensions use a popup to provide functionality, such as displaying a list of tabs, or additional information regarding the current tab. Users can easily find it by clicking on the extension toolbar icon. When the user navigates away it will automatically close.

[The options page][doc-options]
: The options page provides a way for users to customize an extension; such as choosing which sites the extension will run on. Users can access the options page in several ways as described in [Finding the options page][doc-options-view].

Other extension HTML pages include [Chrome override pages][doc-override], [sandbox pages][doc-sandbox] or any custom page included for a specific purpose like onboarding the user.

### Other assets {: #assets }

> An extension can include many types of resources, but only the [extension icons][manifest-icons] are required. All assets, including images and fonts, must be part of the extension package.

Reviewer: > All assets, including images and fonts, must be part of the extension package.

Is this a reference to remotely hosted code restrictions? If so, this isn't quite true. While remote scripts cannot be executed, the use of remotely hosted images, fonts, styles, configuration, and other non-code resources is acceptable.

Author: Thanks for the clarification! I removed this sentence.

## How they work together {: #interact }

In this section, we will describe how these extension components communicate, store data, and share access to resources.

### Sending messages {: #pageComm }

Many times content scripts, or other extension pages, need to send or receive information from the extension service worker. In these cases, either side can listen for messages sent from the other end, and respond on the same channel. Extensions can send a one-time request or establish a long-lived connection to support multiple messages.

See [Message passing][doc-messages] for more details.

### Storing data {: #data }

> Chrome provides extensions with an optimized [Storage API][api-storage] available to all extension components. It includes four separate storage areas for specific use cases and an event listener that tracks whenever data is updated. For example, when you save changes in the popup, the extension service worker can respond with specified logic.

Reviewer: "Optimized" has implications that may not be accurate. It would be safer to say that the Storage API is purpose-built, specialized, or designed for extensions. 

```suggestion
Chrome provides extensions with a specialized [Storage API][api-storage] available to all extension components. It includes four separate storage areas for specific use cases and an event listener that tracks whenever data is updated. For example, when you save changes in the popup, the extension service worker can respond with specified logic.
```

### Web-accesible resources {: #web-resources }
See [Storage API][api-storage] for usage and code samples.

Web-accessible resources are files (images, HTML, CSS, Javascript) inside an extension that can be
accessed by a content script, web pages, or other extensions. 
### Referencing extension resources {: #ref-files }

You can declare which resources are exposed and to what origins in the
manifest:
Extension HTML pages can use the same tags as a regular HTML page to add an extension asset, but all resources must be included in the extension bundle. Content Scripts can also access extension resources but require some extra steps described in [Accessing extension files in Content Scripts][doc-ref].

```json
{
  ...
  "web_accessible_resources": [
    {
      "resources": [ "images/*.png" ],
      "matches": [ "https://example.com/*" ]
    }
  ],
  ...
}
```

See [Web-accesible resources][docs-web-acc-res] for usage information.

> ## Using Chrome APIs {: #apis }

Author: The Chrome API section will be moved to [Reference API](https://developer.chrome.com/docs/extensions/reference/) issue #[4340](https://github.com/GoogleChrome/developer.chrome.com/issues/4340)

This will allow developers to find implementation details about the Chrome APIs in the main [reference docs](https://developer.chrome.com/docs/extensions/reference/)

layout: "layouts/doc-post.njk"
title: "Architecture overview"
date: 2012-09-18
updated: 2022-12-16
description: A high-level explanation of the architecture of Chrome Extensions.
subhead: A high-level explanation of the structure of a Chrome Extension.
anchorRedirects:
  view_page: /docs/extensions/mv3/options/#view_page
  files: /docs/extensions/mv3/content_scripts/#files
---

## Overview {: #overview }

A Chrome extension is composed of different parts. This page describes the structure of an extension, the role each part plays and how they work together. It does not describe the code-level details of how to write an extension.

If you are not familiar with Chrome extension development, we recommend first reading [Extensions 101][doc-ext-101] and [Development Basics][doc-dev-basics].

## The structure of a Chrome extension {: #arch }

### The manifest {: #manifest }

> The manifest is the configuration file of a Chrome extension. It is a required JSON file that must be located at the root of the project. It provides the browser with a blueprint of the extension, with important information such as:

Author: I added a link to [Development basics - Structure](https://developer.chrome.com/docs/extensions/mv3/getstarted/development-basics/#structure). I think this might be sufficient. If not, we can circle back later.

### The service worker {: #background_script }

A service worker is an event-based script that the browser runs in the background. It is often used to process data, coordinate tasks in different parts of an extension, and as an extension's event manager. For example, the service worker can listen for and react to events when the extension is first installed, a new tab is created, a new bookmark is added, the extension toolbar icon is clicked, etc.

A service worker can access all the [Extension APIs][api-ref], but as a type of [Worker][mdn-worker] it can't use the typical DOM APIs that a document's global Window object provides. It also runs in its own environment, so it cannot directly modify a web page's content.

See [Handling events in the extension service worker][doc-sw] for more details. 

### Content scripts {: #content-scripts }

Extensions use content scripts to inject code into host pages. They allow the extension to interact with and modify pages in the browser. For example, they can insert a new element on the page, change the style of a website, modify the [DOM][mdn-dom] elements, etc. 

> Content Scripts share access to the same DOM tree as the host page but run in a separate JavaScript environment. They also have access to a limited number of [Chrome APIs][api-ref]. See [Understanding content scripts][doc-content-scripts] for more details.

Reviewer: Minor change to call out & link to a key terms. Please replace the inline link with a footer link.

```suggestion
Content Scripts share access to the same DOM tree as the host page but run in a separate JavaScript environment (the extension's [isolated world](/docs/extensions/mv3/content_scripts/#isolated_world)). They also have access to a limited number of [Chrome APIs][api-ref]. See [Understanding content scripts][doc-content-scripts] for more details.
```

### Extension HTML pages {: #html-files }

An extension can have different HTML pages depending on the design. All extension HTML files can use the [Chrome APIs][api-ref], but cannot include inline Javascript; they must point to a JavaScript file. The two most common HTML pages are:

[The popup][doc-popup]
: Many extensions use a popup to provide functionality, such as displaying a list of tabs, or additional information regarding the current tab. Users can easily find it by clicking on the extension toolbar icon. When the user navigates away it will automatically close.

[The options page][doc-options]
: The options page provides a way for users to customize an extension; such as choosing which sites the extension will run on. Users can access the options page in several ways as described in [Finding the options page][doc-options-view].

Other extension HTML pages include [Chrome override pages][doc-override], [sandbox pages][doc-sandbox] or any custom page included for a specific purpose like onboarding the user.

### Other assets {: #assets }

> An extension can include many types of resources, such as images and fonts, but only the [extension icons][manifest-icons] are required for extensions hosted in the [Chrome Web Store][cws].

Reviewer: Not required, but it might be good to add another sentence to this paragraph like …

```suggestion
An extension can include many types of resources, such as images and fonts, but only the [extension icons][manifest-icons] are required for extensions hosted in the [Chrome Web Store][cws]. Also, [Chrome Web Store policy](/docs/webstore/program-policies/mv3-requirements/) requires that extensions include all code that the extension executes in the extension's package.
```

Please replace the inline link in this suggestion with a footer link.

layout: "layouts/doc-post.njk"
title: "Architecture overview"
date: 2012-09-18

> updated: 2023-01-09

Author: ```suggestion
updated: 2023-01-10
```

If you are not familiar with Chrome extension development, we recommend first reading [Extensions 101][doc-ext-101] and [Development Basics][doc-dev-basics].

## The structure of a Chrome extension {: #arch }

### The manifest {: #manifest }

The manifest is the configuration file of a Chrome extension. It is a required JSON file that must be located at the [root of the project][dev-basics-structure]. It provides the browser with a blueprint of the extension, with important information such as:

- The name of the extension, a description of what it does, the current version number, and what icons to use.
- The [Chrome API][api-ref] keys and [permissions][doc-perms] that the extension needs.
- The files assigned as the extension service worker, the popup HTML file, the options page, the content scripts, etc.

The [Manifest keys][doc-manifest] article contains the complete list of default and optional keys. For copy-paste-ready code samples, check out the [Manifest examples][doc-manifest-examples].

### The extension service worker {: #background_script }

> An extension service worker is an event-based script that the browser runs in the background. It is often used to process data, coordinate tasks in different parts of an extension, and as an extension's event manager. For example, the service worker can listen for and react to events when the extension is first installed, a new tab is created, a new bookmark is added, the extension toolbar icon is clicked, etc.

Reviewer 2: Where do we mention that service workers don't need to be registered in extension HTML pages? People seem to assume that they are and we had another issue (#4800) opened on this.

Author: @jpmedley I think this is out of scope. We should create a separate PR to address this.

Reviewer 2: I've assigned #4800 to you.
